pkg/app/carbonapi/proc.go

```diff
@@ -0,0 +1,65 @@
+package carbonapi
+
+import (
+	"context"
+	"time"
+
+	"github.com/bookingcom/carbonapi/pkg/carbonapipb"
+)
+
+// ProcessRequests processes the queued requests.
+// TODO: Handler request timeouts. The timed-out requests don't need to be forwarded.
+//       Currently, they'll be handled by the old limiter that's still in place.
+func ProcessRequests(app *App) {
+	// semaphore does what semaphores do: It limits the number of concurrent requests.
+	semaphore := make(chan bool, app.config.MaxConcurrentUpstreamRequests)
```

Comment:
As this is only a semaphore, we can consider changing it to signal only channel by modifying it to `chan struct{}`

Reply:
I knew you'll have this comment) I did do this before, but I think bools are more clean.

Reply:
So, this is intentional. I always prefer bools to `struct{}` in such cases because `struct{}` has a more complicated semantics.